[blur][iOS] Fix issues with blur not blurring text content properly

bbarthec opened the pull request:
# Why

Resolves [ENG-1418](https://linear.app/expo/issue/ENG-1418/fix-outstanding-issues-with-expo-blur-on-ios)
Resolves https://github.com/expo/expo/issues/12871

# How

- [x] I've got rid of the previous blur mechanism that was basing on `alpha` manipulation, because this is [directly discouraged by Apple](https://developer.apple.com/documentation/uikit/uivisualeffectview) and it was not working very well 😞 
- [x] I've tested many available solutions
  - [x] [**react-native-blur**](https://github.com/Kureev/react-native-blur) is using [some undocumented (possibly `private` API](https://github.com/Kureev/react-native-blur/blob/master/ios/BlurEffectWithAmount.m#L43-L62)) and that discouraged me to use it (even though it is actually working)
  - [x] [**bluuur** native library](https://github.com/ML-Works/Bluuur) is using even [more private API](https://github.com/ML-Works/Bluuur/blob/master/Bluuur/Classes/MLWBluuurView.m#L68) and it's stopped working since **iOS 14**
  - [x] [**`UIViewPropertyAnimator`-based Gist**](https://gist.github.com/darrarski/29a2a4515508e385c90b3ffe6f975df7) is a clever way to take advantage of setting it's `fractionComplete` property and that's a solution I've adapted
- [x] I've rewritten almost all native iOS codebase to adapt `UIViewPropertyAnimator`-based approach to achieve blur intensity manipulation
- [x] I've enriched `BlurViewScreen` available in the `ncl` project with more text, fancy random images (fetched from https://source.unsplash.com/) and manual control over intensity value.

# Test Plan

- [x] I've manually tested backgrounding and foregrounding the app as this was mentioned it might cause problems, but it's working flawlessly
- [x] The solution I've proposed seems to be very resource-expensive as it recreates the animator and blur effect upon every render, but I've seen no framedrops in the inspector

## Recording of old behaviour

https://user-images.githubusercontent.com/16623003/139324716-f00d2491-087d-4aed-b6a9-52a868e1477f.mp4

## Recordings of new behaviour

| **iOS@13.7** (simulator) | **iOS@14.8** (actual device) | **iOS@15.0** (simulator) |
|---|---|---|
| <video src="https://user-images.githubusercontent.com/16623003/139319452-0ce82747-b9ca-4a33-96f7-9c9c30c1189e.mp4" /> | <video src="https://user-images.githubusercontent.com/16623003/139318654-2f19e895-5369-44f9-bb63-60ab7b065d97.mp4" /> | <video src="https://user-images.githubusercontent.com/16623003/139319506-131daf97-c560-4ad0-a11b-9d0b244c0fcd.mp4" /> |

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

bbarthec committed:
[iOS][JS][docs] Refactor expo-blur code & remove setNativeProps

bbarthec committed:
[blur][iOS] Convert EXBlurView to Swift

bbarthec committed:
[blur][iOS] Rewrite the module

- removed old Objective-C code that based on the blur mechanism explicitly discouraged by Apple in their docs (it was not working anyway)
- added new Swift code with blur mechanism basing on `UIViewPropertyAnimator`

bbarthec committed:
[blur][iOS] Restore light tint being mapped to .extraLight UIBlurEffect.Style

bbarthec committed:
[ncl][blur] Enrich blur screen

- added more feedback text
- added manual intensity slider

bbarthec committed:
[ncl][blur] Fix iOS issues

bbarthec committed:
[blur] Update CHANGELOG

tsapeta reviewed (commented):
Cool that you also migrated code to Swift 👍

One nitpick: in Swift we don't need to prefix filenames nor symbols with `EX`. The only situation where it is needed is when you use them from Objective-C, but in that case I'd prefer
```swift
@objc(EXBlurView)
public class BlurView: UIView {
```

which exposes `BlurView` class as `EXBlurView` when used in Objective-C.

bbarthec committed:
[blur][iOS] Apply suggested formatting

- Added formatting suggested in review
- Replaced `React.ComponentProps<typeof View>` with `ViewProps`

Simek commented on packages/expo-blur/src/BlurView.tsx:
```diff
@@ -1,41 +1,24 @@
-import { BlurProps, BlurTint, ComponentOrHandle } from './BlurView.types';
+import { BlurProps } from './BlurView.types';
 
 export default class BlurView extends React.Component<BlurProps> {
```
Since most of the class related code has been removed, can we refactor this component to the functional one?

bbarthec replied:
See the current implementation. I've researched the topic and there's no other way to achieve what's I was trying to do (ability to pass `ref={useRef}` prop and properly separate view-based props and blur-based props) 🤷

bbarthec committed:
[blur-view][iOS] Restyle code & add setNativeProps method

- Reformatted code style according to the review comments
- Added back `setNativeProps` in place for removed `updateNativeProps`.
It will be used to animate available props on BlurView, especially `intensity`.
- Removed `UIBlurEffect+Extensions.swift` in favour of inlined static function.

bbarthec committed:
[blur] Renamed `BlurProps` into `BlurViewProps` and fixed/enabled animating any `BlurView`'s prop using `setNativeProps` method

- Renamed `BlurProps` into `BlurViewProps` to match the `react-native`'s types names convention
- Created wrapper around `forwardedRef` mechanism to enable `BlurView`
to be treated as a regular `View` with additional properties.
Additionally added a propoer separation between plain `View`'s props and
actual `BlurView`'s props.

bbarthec committed:
[blur][docs] Regenerate expo-blur.json

bbarthec committed:
[blur][ncl] Update blurView demo screen

bbarthec commented on packages/expo-blur/ios/EXBlur/EXBlurViewManager.m:
```diff
@@ -31,39 +31,11 @@ - (NSString *)viewName
-  [view didSetProps:@[@"intensity"]];
-}
-
-EX_EXPORT_METHOD_AS(updateProps,
```
Yeah, I've found out the undocumented usage. I've restored the mechanism and even fixed it to work more properly 😅

bbarthec committed:
[blur][iOS] replace Float with Double type for `intensity` prop

bbarthec committed:
[blur] Fix nullability issues

tsapeta commented on packages/expo-blur/src/index.ts:
```diff
@@ -1,2 +1,2 @@
 export { default as BlurView } from './BlurView';
-export { BlurTint, BlurProps } from './BlurView.types';
+export { BlurTint, BlurViewProps } from './BlurView.types';
```
Isn't it a breaking change if someone imported this type? 🤔

bbarthec replied:
Thank you for catching this! I intended to add this change to the CHANGELOG, but forgot about this later on 😅

bbarthec committed:
[blur][skip ci] Added missing CHANGELOG entry
